File: go/robot-simulator/robot_simulator.go
package robot

const testVersion = 3

// define directions
const (
	N Dir = iota
	E
	S
	W
)

// Right turns Step1Robot to the right
func Right() {
	Step1Robot.Dir = (Step1Robot.Dir + 1) % 4
}

// Left turns Step1Robot to the left
func Left() {
	Step1Robot.Dir = (Step1Robot.Dir + 3) % 4
}

// Advance moves Step1Robot one position forward in the direction it's facing
func Advance() {
	switch Step1Robot.Dir {
	case N:
		Step1Robot.Y++
	case S:
		Step1Robot.Y--
	case E:
		Step1Robot.X++
	case W:
		Step1Robot.X--
	}
}

// String converts a direction to a string representation
func (direction Dir) String() (s string) {
	switch direction {
	case N:
		s = "North"
	case S:
		s = "South"
	case E:
		s = "East"
	case W:
		s = "West"
	}

	return
}

File: clojure/robot-simulator/src/robot_simulator.clj
(ns robot-simulator)

(defn robot
  "Initialize a robot."
  [coordinates bearing]
  (assoc {} :bearing bearing :coordinates coordinates)
  )

(def dx {:north 0, :south 0, :east 1, :west -1})
(def dy {:north 1, :south -1, :east 0, :west 0})

(defn- advance
  "Move robot forward one position in current direction."
  [{bearing :bearing {x :x y :y} :coordinates}]
    (robot {:x (+ x (dx bearing)), :y (+ y (dy bearing))} bearing))

(def left {:north :west, :south :east, :east :north, :west :south})
(def right {:north :east, :south :west, :east :south, :west :north})

(defn- turn
  "Turn robot left or right."
  [direction {bearing :bearing coordinates :coordinates}]
  (robot coordinates (direction bearing)))

(def instruction {\A advance, \L (partial turn left), \R (partial turn right)})

(defn- execute
  "Execute a single instruction."
  [robot i]
  ((instruction i) robot))

(defn simulate
  "Execute a series of instructions, returning the resulting robot."
  [instructions robot]
  (reduce execute robot instructions))
